src/pydvl/value/shapley/montecarlo.py

```diff
@@ -86,123 +216,139 @@ def truncated_montecarlo_shapley(
+            truncation=truncation,
             worker_id=worker_id,
-            progress=progress,
             update_period=worker_update_period,
             config=config,
         )
         for worker_id in range(n_jobs)
     ]
-    for worker_id in range(n_jobs):
-        workers[worker_id].run(block=False)
+    for worker in workers:
+        worker.run(block=False)
 
-    while not coordinator.check_done():
+    while not coordinator.check_convergence():
         sleep(coordinator_update_period)
 
-    values, stderr = coordinator.get_results()
+    return coordinator.accumulate()
 
-    return ValuationResult(
-        algorithm="truncated_montecarlo_shapley",
-        status=coordinator.status(),
-        values=values,
-        stderr=stderr,
-        data_names=u.data.data_names,
-    )
+    # Something like this would be nicer, but it doesn't seem to be possible
+    # to start the workers from the coordinator.
+    # coordinator.add_workers(
+    #     n_workers=n_jobs,
+    #     u=u_id,
+    #     update_period=worker_update_period,
+    #     config=config,
+    #     truncation=truncation,
+    # )
+    #
+    # return coordinator.run(delay=coordinator_update_period)
 
 
-def _permutation_montecarlo_marginals(
-    u: Utility, max_permutations: int, progress: bool = False, job_id: int = 1
-) -> "NDArray":
+def _permutation_montecarlo_shapley(
+    u: Utility,
+    *,
+    done: StoppingCriterion,
+    truncation: TruncationPolicy,
+    algorithm_name: str = "permutation_montecarlo_shapley",
+    progress: bool = False,
+    job_id: int = 1,
+) -> ValuationResult:
     """Helper function for :func:`permutation_montecarlo_shapley`.
 
     Computes marginal utilities of each training sample in
     :obj:`pydvl.utils.utility.Utility.data` by iterating through randomly
     sampled permutations.
 
     :param u: Utility object with model, data, and scoring function
-    :param max_permutations: total number of permutations to use
+    :param done: Check on the results which decides when to stop
+    :param truncation: A callable which decides whether to interrupt
+        processing a permutation and set all subsequent marginals to zero.
+    :param algorithm_name: For the results object. Used internally by different
+        variants of Shapley using this subroutine
     :param progress: Whether to display progress bars for each job.
     :param job_id: id to use for reporting progress (e.g. to place progres bars)
-    :return: a matrix with each row being a different permutation and each
-        column being the score of a different data point
+    :return: An object with the results
     """
-    n = len(u.data)
-    values = np.zeros(shape=(max_permutations, n))
-    pbar = maybe_progress(max_permutations, progress, position=job_id)
-    for iter_idx in pbar:
+    result = ValuationResult.empty(algorithm=algorithm_name, indices=u.data.indices)
+
+    pbar = tqdm(disable=not progress, position=job_id, total=100, unit="%")
```

Comment:
Not using `maybe_progress`, but still ugly. We need a one-liner that updates a percentage